File: 0002-even_fibonacci_numbers/main.py
#!/usr/bin/env python

from typing import Generator

def fib(limit: int) -> Generator[int, None, None]:
    last = 0
    current = 1

    while True:
        new = last + current

        if new > limit:
            break

        yield new

        last = current
        current = new


def main():
    limit = 4_000_000
    result = sum(
        number for number in fib(limit)
        if number % 2 == 0
    )
    print(result)


if __name__ == "__main__":
    main()

File: 0002-even_fibonacci_numbers/main.rs
fn fib(limit: i32, mut last: i32) -> Vec<i32> {
    let mut result: Vec<i32> = vec!();
    let mut current: i32 = 1;
    
    loop {
        let new: i32 = current + last;        
        
        if new > limit {
            break;
        }
        
        result.push(new);
        last = current;
        current = new;
    }
    
    result
}

fn main() {
    let limit: i32 = 4_000_000;
    let result: i32 = fib(limit, 0)
        .into_iter()
        .filter(|n| n % 2 == 0)
        .sum();
        
    println!("{}", result);
}
